sampler.go

```diff
@@ -174,6 +181,177 @@ func (s *rateLimitingSampler) Equal(other Sampler) bool {
 
 // -----------------------
 
+type lowerBoundSampler struct {
+	rateLimitingSampler Sampler
+	samplingRate        float64
+	tags                []Tag
+}
+
+// NewLowerBoundSampler creates a sampler that samples at most maxTracesPerSecond. It is a wrapper around
+// a rateLimitingSampler with different tags since it's intended to be used by the adaptiveSampler.
+// The sampler.param tag is sampling rate rather than maxTracesPerSecond. This is used to tell the collector
+// that the trace sampled by this sampler used the samplingRate to determine the throughput for the operation.
+func NewLowerBoundSampler(maxTracesPerSecond, samplingRate float64) (Sampler, error) {
+	tags := []Tag{
+		{key: SamplerTypeTagKey, value: SamplerTypeLowerBound},
+		{key: SamplerParamTagKey, value: samplingRate},
+	}
+	rateLimitingSampler, err := NewRateLimitingSampler(maxTracesPerSecond)
+	if err != nil {
+		return nil, err
+	}
+	return &lowerBoundSampler{
+		rateLimitingSampler: rateLimitingSampler,
+		samplingRate:        samplingRate,
+		tags:                tags,
+	}, nil
+}
+
+// IsSampled implements IsSampled() of Sampler.
+func (s *lowerBoundSampler) IsSampled(id uint64, operation string) (bool, []Tag) {
+	sampled, _ := s.rateLimitingSampler.IsSampled(id, operation)
+	return sampled, s.tags
+}
+
+func (s *lowerBoundSampler) Close() {
+	s.rateLimitingSampler.Close()
+}
+
+func (s *lowerBoundSampler) Equal(other Sampler) bool {
+	if o, ok := other.(*lowerBoundSampler); ok {
+		return s.samplingRate == o.samplingRate &&
+			s.rateLimitingSampler.Equal(o.rateLimitingSampler)
+	}
+	return false
+}
+
+// -----------------------
+
+// sampler for the operation level
+type operationSampler struct {
+	probabilisticSampler Sampler
+	lowerBoundSampler    Sampler
+	operation            string
+}
+
+// NewOperationSampler operationSampler is a delegating sampler that applies both probabilisticSampler and
+// rateLimitingSampler. The probabilisticSampler is given higher priority when tags are emitted,
+// ie. if IsSampled() for both samplers return true, the tags for probabilisticSampler will be used.
+func NewOperationSampler(operation string, maxTracesPerSecond, samplingRate float64) (Sampler, error) {
+	probabilisticSampler, err := NewProbabilisticSampler(samplingRate)
+	if err != nil {
+		return nil, err
+	}
+	lowerBoundSampler, err := NewLowerBoundSampler(maxTracesPerSecond, samplingRate)
+	if err != nil {
+		return nil, err
+	}
+	return &operationSampler{
+		probabilisticSampler: probabilisticSampler,
+		lowerBoundSampler:    lowerBoundSampler,
+		operation:            operation,
+	}, nil
+}
+
+func (s *operationSampler) IsSampled(id uint64, operation string) (bool, []Tag) {
+	if sampled, tags := s.probabilisticSampler.IsSampled(id, operation); sampled {
+		return true, tags
+	}
+	return s.lowerBoundSampler.IsSampled(id, operation)
```

Comment:
I suggest replacing `lower bound` here with a standard rate limiting sampler, but overriding the returned tag values by using the probability of the probabilisticSampler